### PR TITLE
Hierarchy builder: Custom nodes support

### PR DESCRIPTION
**Pull request by grigasp**
Closes https://github.com/iTwin/presentation/issues/238

### CODE REVIEW

File: packages/hierarchy-builder/src/test/internal/operators/MergeInstanceNodesByLabel.test.ts [hunk 1]
```diff
@@ -26,35 +33,68 @@ describe("mergeInstanceNodesByLabelOperator", () => {
...
-      createTestNode({ key: { type: "instances", instanceKeys: [createTestInstanceKey({ id: "0x1" })] }, mergeByLabelId: "" }),
-      createTestNode({ key: { type: "instances", instanceKeys: [createTestInstanceKey({ id: "0x2" })] }, mergeByLabelId: "" }),
+      createTestNode({ key: { type: "instances", instanceKeys: [createTestInstanceKey({ id: "0x1" })] }, params: { mergeByLabelId: "" } }),
+      createTestNode({ key: { type: "instances", instanceKeys: [createTestInstanceKey({ id: "0x2" })] }, params: { mergeByLabelId: "" } }),
     ];
     const result = await getObservableResult(from(nodes).pipe(createMergeInstanceNodesByLabelOperator(directNodesCache)));
     expect(result).to.deep.eq(nodes);
   });
 
   it("doesnt merge nodes that have different `mergeByLabelId`", async () => {
```
Comment (JonasDov) [inline packages/hierarchy-builder/src/test/internal/operators/MergeInstanceNodesByLabel.test.ts:43]:
```suggestion
  it("doesn't merge nodes that have different `mergeByLabelId`", async () => {
```

File: packages/hierarchy-builder/src/test/internal/operators/MergeInstanceNodesByLabel.test.ts [hunk 2]
```diff
@@ -26,35 +33,68 @@ describe("mergeInstanceNodesByLabelOperator", () => {
 
   it("doesnt merge nodes that have empty `mergeByLabelId`", async () => {
```
Comment (JonasDov) [inline packages/hierarchy-builder/src/test/internal/operators/MergeInstanceNodesByLabel.test.ts:34]:
```suggestion
  it("doesn't merge nodes that have empty `mergeByLabelId`", async () => {
```

File: packages/hierarchy-builder/src/test/internal/operators/MergeInstanceNodesByLabel.test.ts [hunk 3]
```diff
@@ -26,35 +33,68 @@ describe("mergeInstanceNodesByLabelOperator", () => {
...
-      createTestNode({ key: { type: "instances", instanceKeys: [createTestInstanceKey({ id: "0x2" })] }, mergeByLabelId: "b" }),
+      createTestNode({ key: { type: "instances", instanceKeys: [createTestInstanceKey({ id: "0x1" })] }, params: { mergeByLabelId: "a" } }),
+      createTestNode({ key: { type: "instances", instanceKeys: [createTestInstanceKey({ id: "0x2" })] }, params: { mergeByLabelId: "b" } }),
     ];
     const result = await getObservableResult(from(nodes).pipe(createMergeInstanceNodesByLabelOperator(directNodesCache)));
     expect(result).to.deep.eq(nodes);
   });
 
-  it("merges nodes that have same `mergeByLabelId`", async () => {
+  it("doesnt merge nodes that have different labels", async () => {
```
Comment (JonasDov) [inline packages/hierarchy-builder/src/test/internal/operators/MergeInstanceNodesByLabel.test.ts:52]:
```suggestion
  it("doesn't merge nodes that have different labels", async () => {
```

File: packages/hierarchy-builder/src/test/internal/operators/ClassGrouping.test.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+      label: classLabel,
+    } as unknown as ECClass);
+    return {
+      id: Id64.invalid,
+      name: fullName,
+      label: classLabel ?? className,
+    };
+  }
+
+  it("doesnt group non-instance nodes", async () => {
```
Comment (JonasDov) [inline packages/hierarchy-builder/src/test/internal/operators/ClassGrouping.test.ts:43]:
```suggestion
  it("doesn't group non-instance nodes", async () => {
```

File: packages/hierarchy-builder/src/hierarchy-builder/internal/operators/HideIfNoChildren.ts [hunk 5]
```diff
@@ -2,46 +2,64 @@
...
               defer(() => {
-                enableLogging && console.log(`HideIfNoChildrenOperator: requesting children flag for ${n.label}`);
+                doLog(`requesting children flag for ${n.label}`);
                 return hasNodes(n).pipe(
-                  map((children) => {
-                    enableLogging && console.log(`HideIfNoChildrenOperator: children for ${n.label}: ${children}`);
-                    return { ...n, children };
-                  }),
+                  log((children) => `determined children for ${n.label}: ${children}`),
+                  map((children) => ({ ...n, children })),
```
[the remark this answers is not in the document]
Comment (grigasp) [reply packages/hierarchy-builder/src/hierarchy-builder/internal/operators/HideIfNoChildren.ts:46]: cleaned up

File: packages/hierarchy-builder/src/test/internal/operators/MergeInstanceNodesByLabel.test.ts [hunk 6]
```diff
@@ -26,35 +33,68 @@ describe("mergeInstanceNodesByLabelOperator", () => {
...
-      createTestNode({ key: { type: "instances", instanceKeys: [createTestInstanceKey({ id: "0x1" })] }, mergeByLabelId: "x" }),
-      createTestNode({ key: { type: "instances", instanceKeys: [createTestInstanceKey({ id: "0x2" })] }, mergeByLabelId: "x" }),
+      createTestNode({ key: { type: "instances", instanceKeys: [createTestInstanceKey({ id: "0x1" })] }, label: "a", params: { mergeByLabelId: "x" } }),
+      createTestNode({ key: { type: "instances", instanceKeys: [createTestInstanceKey({ id: "0x2" })] }, label: "b", params: { mergeByLabelId: "x" } }),
+    ];
+    const result = await getObservableResult(from(nodes).pipe(createMergeInstanceNodesByLabelOperator(directNodesCache)));
+    expect(result).to.deep.eq(nodes);
+  });
+
+  it("doesnt merge nodes of different types", async () => {
```
[the remark this answers is not in the document]
Comment (grigasp) [reply packages/hierarchy-builder/src/test/internal/operators/MergeInstanceNodesByLabel.test.ts]: fixed

File: packages/hierarchy-builder/src/test/internal/operators/MergeInstanceNodesByLabel.test.ts (hunk 1 above)
Comment (grigasp) [reply packages/hierarchy-builder/src/test/internal/operators/MergeInstanceNodesByLabel.test.ts]: fixed

File: packages/hierarchy-builder/src/test/internal/operators/MergeInstanceNodesByLabel.test.ts (hunk 2 above)
Comment (grigasp) [reply packages/hierarchy-builder/src/test/internal/operators/MergeInstanceNodesByLabel.test.ts]: fixed

File: packages/hierarchy-builder/src/test/internal/operators/MergeInstanceNodesByLabel.test.ts (hunk 3 above)
Comment (grigasp) [reply packages/hierarchy-builder/src/test/internal/operators/MergeInstanceNodesByLabel.test.ts]: fixed

File: packages/hierarchy-builder/src/test/internal/operators/ClassGrouping.test.ts (hunk 4 above)
Comment (grigasp) [reply packages/hierarchy-builder/src/test/internal/operators/ClassGrouping.test.ts]: fixed

**Review by JonasDov — APPROVED** [review]

**Pull request merged**